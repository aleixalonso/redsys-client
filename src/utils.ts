import * as crypto from "crypto";

export function zeroPad(smth: string, blocksize: number) {
  const buf = Buffer.from(smth.toString(), "utf8");
  const pad = Buffer.alloc(
    (blocksize - (buf.length % blocksize)) % blocksize,
    0
  );

  return Buffer.concat([buf, pad]);
}

export function encrypt3DES(orderId: string, secret: string) {
  const secretKey = Buffer.from(secret, "base64");
  const iv = Buffer.alloc(8, 0);
  const cipher = crypto.createCipheriv("des-ede3-cbc", secretKey, iv);
  cipher.setAutoPadding(false);
  const paddedStr = zeroPad(orderId, 8);

  return (
    cipher.update(paddedStr.toString(), "utf8", "base64") +
    cipher.final("base64")
  );
}

export function mac256(orderEncoded: string, merchantData: string) {
  const hexMac256 = crypto
    .createHmac("sha256", Buffer.from(orderEncoded, "base64"))
    .update(merchantData)
    .digest("hex");

  return Buffer.from(hexMac256, "hex").toString("base64");
}

// Convert to URL-safe Base64
export function fromBase64(base64: string): string {
  return base64
    .replace(/\+/g, "-") // Replace + with -
    .replace(/\//g, "_") // Replace / with _
    .replace(/=+$/, ""); // Remove padding characters
}
