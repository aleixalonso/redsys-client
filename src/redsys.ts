import { MerchantParameters } from "./types";
import { encrypt3DES, fromBase64, mac256 } from "./utils";

export class Redsys {
  private _merchantParameters: MerchantParameters;

  constructor() {
    this._merchantParameters = {
      Ds_Merchant_MerchantCode: "",
      Ds_Merchant_Terminal: "",
      Ds_Merchant_TransactionType: "",
      Ds_Merchant_Amount: 0,
      Ds_Merchant_Currency: "",
      Ds_Merchant_Order: "",
      Ds_Merchant_MerchantURL: "",
    };
  }

  // setters and getters

  public set merchantParameters(merchantParameters: MerchantParameters) {
    this._merchantParameters = merchantParameters;
  }

  public get merchantParameters(): MerchantParameters {
    return this._merchantParameters;
  }

  // public methods

  public createMerchantParameters(): string {
    return Buffer.from(JSON.stringify(this.merchantParameters)).toString(
      "base64"
    );
  }

  public createMerchantSignature(secret: string): string {
    const merchantParameters = this.createMerchantParameters();
    // Encrypt order
    const encodedOrder = encrypt3DES(
      this.merchantParameters.Ds_Merchant_Order,
      secret
    );
    // Calculate signature
    const signature = mac256(encodedOrder, merchantParameters);

    return signature;
  }

  public createMerchantSignatureNotif(secret: string, params: string): string {
    const decodedData = JSON.parse(Buffer.from(params, "base64").toString());

    // Encrypt order
    const orderEncoded = encrypt3DES(decodedData.Ds_Order, secret);

    // Compute signature with the order encoded and params from response
    const computedSignature = mac256(orderEncoded, params);

    // Return the signature in URL-safe Base64
    return fromBase64(computedSignature);
  }
}
