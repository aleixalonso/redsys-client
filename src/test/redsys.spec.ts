import { Redsys } from "../redsys";
import { fromBase64 } from "../utils";

describe("Redsys", () => {
  let redsys: Redsys;

  beforeEach(() => {
    redsys = new Redsys();
    redsys.merchantParameters = {
      Ds_Merchant_MerchantCode: "999008881",
      Ds_Merchant_Terminal: "1",
      Ds_Merchant_TransactionType: "0",
      Ds_Merchant_Amount: 1000,
      Ds_Merchant_Currency: "978",
      Ds_Merchant_Order: "1234567890",
      Ds_Merchant_MerchantURL: "https://www.example.com",
    };
  });

  describe("createMerchantParameters", () => {
    it("Should the expected base64 string", () => {
      const result = redsys.createMerchantParameters();
      expect(result).toMatchSnapshot();
    });
  });

  describe("createMerchantSignature", () => {
    it("Should the expected signature", () => {
      const result = redsys.createMerchantSignature(
        "sq7HjrUOBfKmC576ILgskD5srU870gJ7" // redsys test secret
      );
      expect(result).toMatchSnapshot();
    });
  });

  describe("createMerchantSignatureNotif", () => {
    it("Should the expected signature", () => {
      // Example params and signature from an ok redirect
      const params =
        "eyJEc19EYXRlIjoiMDYlMkYwNyUyRjIwMjQiLCJEc19Ib3VyIjoiMTglM0EwMSIsIkRzX1NlY3VyZVBheW1lbnQiOiIxIiwiRHNfQW1vdW50IjoiMTAwMCIsIkRzX0N1cnJlbmN5IjoiOTc4IiwiRHNfT3JkZXIiOiIxNzIwMjgxNjMxNTUiLCJEc19NZXJjaGFudENvZGUiOiI5OTkwMDg4ODEiLCJEc19UZXJtaW5hbCI6IjAwMSIsIkRzX1Jlc3BvbnNlIjoiMDAwMCIsIkRzX1RyYW5zYWN0aW9uVHlwZSI6IjAiLCJEc19NZXJjaGFudERhdGEiOiIiLCJEc19BdXRob3Jpc2F0aW9uQ29kZSI6IjA5NDc3OSIsIkRzX0NhcmRfTnVtYmVyIjoiNDU0ODgxKioqKioqMDAwNCIsIkRzX0NvbnN1bWVyTGFuZ3VhZ2UiOiIxIiwiRHNfQ2FyZF9Db3VudHJ5IjoiNzI0IiwiRHNfQ2FyZF9CcmFuZCI6IjEiLCJEc19Qcm9jZXNzZWRQYXlNZXRob2QiOiI3OCIsIkRzX0VDSSI6IjA1IiwiRHNfUmVzcG9uc2VfRGVzY3JpcHRpb24iOiJPUEVSQUNJT04rQVVUT1JJWkFEQSIsIkRzX0NvbnRyb2xfMTcyMDI4MTY2MzMxNSI6IjE3MjAyODE2NjMzMTUifQ==";
      const receivedSignature = "VF8kC7KJlVCn45Gn_75MEB2dZYsjOs5QZHrOValj5BU=";

      const computedSignature = redsys.createMerchantSignatureNotif(
        "sq7HjrUOBfKmC576ILgskD5srU870gJ7", // redsys test secret
        params
      );

      expect(computedSignature).toMatchSnapshot();
    });
  });
});
