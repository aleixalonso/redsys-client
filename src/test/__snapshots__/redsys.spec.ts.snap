// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redsys createMerchantParameters Should the expected base64 string 1`] = `"eyJEc19NZXJjaGFudF9NZXJjaGFudENvZGUiOiI5OTkwMDg4ODEiLCJEc19NZXJjaGFudF9UZXJtaW5hbCI6IjEiLCJEc19NZXJjaGFudF9UcmFuc2FjdGlvblR5cGUiOiIwIiwiRHNfTWVyY2hhbnRfQW1vdW50IjoxMDAwLCJEc19NZXJjaGFudF9DdXJyZW5jeSI6Ijk3OCIsIkRzX01lcmNoYW50X09yZGVyIjoiMTIzNDU2Nzg5MCIsIkRzX01lcmNoYW50X01lcmNoYW50VVJMIjoiaHR0cHM6Ly93d3cuZXhhbXBsZS5jb20ifQ=="`;

exports[`Redsys createMerchantSignature Should the expected signature 1`] = `"OQjbr1NseO5mhg/l7nq6a5e6LOpiQ1BpSv+sfz+6/e0="`;
